### PR TITLE
MQTT intitial implementation

### PR DESCRIPTION
**Pull request by gianricod**
hi @cedricve  
this is an initial implementation of enhancement #89.

json data sent to the topic does not make much sense at the moment  and use cases are needed to understand what would be better to send.

i will update documentation to include this io device as soon as i can, with info about the settings.

libmosquittopp is required to be installed . i have modified cmake file to detect that.
documentation for manual install/compilation should also report that.

### CODE REVIEW

**Comment by cedricve** [comment]
amazing contribute @gianricod, let me review your work! Awesome.

**Comment by cedricve** [comment]
@gianricod overall looks good! Styling and implementation is perfect! I was just wondering if we can compile the mosquito library together. For example this is how we do with other dependencies. https://github.com/kerberos-io/machinery/blob/develop/cmake/External-Restclient.cmake

What do you think?

Cédric

**Comment by gianricod** [comment]
Hi @cedricve 
It sounds reasonable and  inline with what you have done so far, and the number of external lib for now is still under control for static linking.
Will modify and test it as soon as i can.

gianrico

**Comment by cedricve** [comment]
hey @gianricod, I'll try to add the cmake file to your branch.

**Comment by gianricod** [comment]
hi cedric
i could not spent too much time on this.   i have started looking at the
lib/cpp makefile of mosquitto and it seemed to lack a .a target

Il giorno gio 2 nov 2017 alle 12:59 Cédric Verstraeten <
notifications@github.com> ha scritto:

> hey @gianricod <https://github.com/gianricod>, I'll try to add the cmake
> file to your branch.
>
-- 

Gianrico D'Angelis

**Comment by gianricod** [comment]
hi @cedricve 
i could spend  time yesterday on it and  i have made some progress.
let me know if you are still working on the cmake file as i have one already and I do not want to risk overlapping.
i have the new  changes ready to be pushed.

gianrico

**Comment by cedricve** [comment]
awesome @gianricod ! I'm trying to build on OSX but I'm getting an error. Any ideas?
    ld: unknown option: --version-script=linker.version

**Comment by gianricod** [comment]
hi @cedricve 
i do not have a mac so i cannot help much as i could not test it.

the problem is that we clone mosquitto and we need to apply a patch to the config.mk to remove : 

-Wl,  --version-script=linker.version

it could be done by using  PATCH_COMMAND and something like 'sed' ( if available on Mac)

for patching with 'patch' we should need to have the patch file somewhere or alternatively something like below (forged for linux, for mac i do not know):

PATCH_COMMAND bash "-c" "/bin/echo 'H4sIAL77AVoAAzM0NUk2NDXhslHg9PF0ivdxcfNxdA+2slXRQOJqKuiG5+jo6palFhVn5ufpFicXZRaU2OZk5mWnFulBRSFqivPzEnNTdXIyk3LziwtLM0tK8vWK8/VUNIL9w1yDgj39/TS5dHV1ueyIsI8YswBmSNHVwAAAAA==' | /usr/bin/base64 -d | /bin/gzip -d | /usr/bin/patch  -p0 config.mk"

Basically the patch is shipped  gzipped and then base64encoded .

**Comment by cedricve** [comment]
Hey @gianricod, sorry for my late answer. I've been playing with your PR and it's great, it works! As mentioned I had some issues on Mac OSX, due to using **make**. What I've did is refactored the External-Project file, so it uses cmake instead of make (I found this in the mosquitto documentation). 

Also for the static library, this feature is missing in the master branch, but I've found out this is already included in the develop branch. Therefore I've used a tag instead of master. Can you take a look and review it? If you're ok, can you modify the External-Project file in your PR, so I can approve your request. I would love to integrate your feature in the next Kerberos.io release.

https://github.com/kerberos-io/machinery/blob/pr/106/cmake/External-Mosquitto.cmake

**Comment by gianricod** [comment]
Hi @cedricve 

i have added the new .cmake file as you see . It looks much better and smoother than the older. 
The only thing is that the clients and broker are also built but it does not add too much time to the compilation so it should be fine.

thanks
gianrico

**Comment by cedricve** [comment]
Awesome! Thanks @gianricod, this will be merged in next release. I already implemented the MQTT settings interface in the web repository today. https://github.com/kerberos-io/web/commit/384db2b03ed68aeb303066f9a6811af2aca2b838

**Comment by flaviostutz** [comment]
About the use cases I have one that I really need:
- I want to detect if a certain motorcycle has passed in front of the camera só that I can do some analytics on this info
- For this task, I want to classify the image by using TensorFlow/Machine Learning. I have this algorithm in a Docker Container
- So I wish I could use MQTT so that if I have multiple classifiers I could distribute the processing among hosts and use specialized containers for each task
- In this scenario, I wish the MQTT message to have:
      - timestamp of image
      - URL for full scale image
      - movement detection intensity
      - option to use masked region/non masked image
      - option to publish the entire image data embedded to the mqtt message (so that my algo doesn’t need to fetch a URL for processing avoiding latency)
      - option to specify a custom topic name, qos and mqtt server

**Comment by cedricve** [comment]
thanks for your feedback @flaviostutz , I think this PR will full fill your requirements. I will try  to publish a new release this weekend.

@gianricod would you be able to create some documentation for the MQTT Io Device? What you need to do is:

* modify this file https://github.com/kerberos-io/documentation/blob/2.0/72_machinery/91_output.md, so the MQTT io device is included.
* add a new file https://github.com/kerberos-io/documentation/tree/2.0/72_machinery/84_Outputs, and add documentation to describe the different parameters and maybe a usecase.

Thanks @gianricod
